### PR TITLE
Support global config with load config package

### PR DESCRIPTION
**Pull request by ertrzyiks**
It's going to address #88 , but it may require more attention in the early phase. Any feedback is welcome.

### CODE REVIEW

**Comment by ai** [comment]
Great! Fast work :). Sorry, I could review it only tomorrow (I am not so fast :( )

**Comment by ertrzyiks** [comment]
Awesome! Please keep in mind that I've just started and submitted PR already because it will not be one liner and any tip on this stage can prevent a headache later on :)

**Comment by ai** [comment]
Looks good. Only few very small things:

1. Why you comment `postcssrc`?
2. Could you please fix my weird manually indent in `package.json` :D

**Comment by ertrzyiks** [comment]
1. In the middle of work my gulp process was dying without any notice. It was fine when I commented out that part. It may even work right now, a lot of code changed since then. I will check it today.
2. Ok, I will revert all unrelated changes there, will leave only a new dependency entry.

From what I see it may be tricky to support pack feature when config loader is used. Are you ok with having it only for the current behavior?

**Comment by ai** [comment]
Feel free to return error on using pack with config

File: index.js [hunk 1]
```diff
@@ -19,89 +30,141 @@ function PostCSSLoaderError(error) {
...
+
+    return plugins;
+}
+
+function loadConfigurationFromRc(callback) {
+    getPostCssRc()
+      .then(function (config) {
+          callback(null, { options: config.options, plugins: config.plugins });
+      })
+      .catch(function () {
```
Comment (ertrzyiks) [inline index.js]: Too optimistic error handling

File: index.js [hunk 2]
```diff
@@ -19,89 +30,141 @@ function PostCSSLoaderError(error) {
...
+      });
+}
 
-    if ( typeof map === 'string' ) map = JSON.parse(map);
-    if ( map && map.mappings ) opts.map.prev = map;
+function getConfiguration(options, pack, callback) {
+    // Read config from separate file or package.json
+    // only if it's not provided explicitly
+    if ( typeof options === 'undefined' ) {
+        if ( pack ) {
```
Comment (ertrzyiks) [inline index.js]: Add spec for that

File: index.js [hunk 3]
```diff
@@ -19,89 +30,141 @@ function PostCSSLoaderError(error) {
...
-            });
-            callback(null, result.css, result.map ? result.map.toJSON() : null);
-            return null;
-        })
-        .catch(function (error) {
-            if ( error.name === 'CssSyntaxError' ) {
-                callback(new PostCSSLoaderError(error));
-            } else {
-                callback(error);
+        var opts  = Object.assign({}, config.options, {
```
Comment (ertrzyiks) [inline index.js]: Replace `Object.assign` with node 0.12 compatible solution

Comment (ertrzyiks) [inline index.js]: Add spec which fails because of `postcss-load-config`, it uses it as well.

**Comment by ertrzyiks** [comment]
@ai I think more or less it's done. I've tried to utilize `postcss-load-config` better, but ended up using it only when there is no explicit `postcss` config for webpack. 

I've added more tests and they require separate configs, so we have multiple gulp tasks generated per each config scenario.

What do you think about it?

**Comment by ai** [comment]
Looks good. Sorry, it could take a while (need be sure, that it will help in webpack 2 too)

**Comment by ai** [comment]
New webpack 2 issue bring new question. Could we do everything with this config?

How we could cover `addDependecyTo` case in this PostCSS global config? https://github.com/postcss/postcss-loader#with-postcss-import

/cc @michael-ciniawsky

**Comment by ertrzyiks** [comment]
I don't think so we can cover `addDependecyTo` in the current version of this config, we have no access to that `webpack` reference when defining configuration. Even with webpack 1 we can not support pack feature, so I would see `load-config` solution as a secondary right now.

**Comment by ai** [comment]
@ertrzyiks do we have any way to pass `webpack` instance to config?

If not, we could ask @michael-ciniawsky for some solution

**Comment by ertrzyiks** [comment]
@ai we could do this but by breaking the whole idea of having config object with primitive values (as it requires a config value being a function). 

Maybe @michael-ciniawsky can come up with some idea. I'm wondering if it would be possible to feed postcss with webpack instance and use internal pub-sub communication between postcss and plugins to propagate it.

**Comment by ai** [comment]
Hm. Maybe we could use `result.message` and set new files to some `dependencies`. And then it will work out of box in postcss-loader?

**Comment by ai** [comment]
@swernerx @MoOx @TrySound @borodean what do you think if all import plugin should set absolute path for all used resources to use it then in webpack?

```js
result.messages.push({ type: 'dependency', path: '/home/ai/test/images/logo.png' })
```

As result hot reload will work out-of-box in postcss-loader, without painful config.

**Comment by borodean** [comment]
@ai that should be fine for postcss-assets, however, I'd need someone's help implementing this.

**Comment by swernerx** [comment]
@ai IMHO this seems to be a good idea. Makes it far easier to track paths from different "viewpoints" and unify them.

**Comment by MoOx** [comment]
Since import plugins are used first, there is no need to put all resources, because the work will be done by webpack when css is inlined by the plugin. So maybe ok for the path, but no need for all resources. Anyway, for postcss-import, there is a total mess to handle. @TrySound made some huge changes and never really finished the job, so there are tons of issues related to race conditions... But current algorithm looks to complicated for my brain, and I am busy with new projects, so no time to handle this... That's why I called for contributor for postcss-import, but, as usual, not a single person responded.

**Comment by michael-ciniawsky** [comment]
@ai @ertrzyiks  I'm on it (addDependencyTo + webpack 2), but I'm on my way to Rome picking up lost ones 👅 , so i will not made it this weekend.Next week I will finish postcss-load-config + gulp + cli, promised 👍 .

@ai  👍 interesting
```js
result.messages.push({ type: 'dependency', path: '/home/ai/test/images/logo.png' })
```
but maybe it's own 'prop'
```
result.file = { root: dirname(options.from), name:  basename(options.from), paths: [] }
```
something like that 😛 

@MoOx  I will take a look at postcss-import after the weekend, basically if I can grasp whats going on. I'm eager to help maintainig it in the future, but must get used to the code base / issues first :).

**Comment by ai** [comment]
@michael-ciniawsky `result.file` requires PostCSS release :). We specially have `messages` for plugins communications.

**Comment by michael-ciniawsky** [comment]
> result.file requires PostCSS release :). We specially have messages for plugins communications.

@ai yep, yep just trolling around 😛, sry for the delay, working on postcss-load-* now. For the webpack issue currently testing if

```js
if (!typeof config === 'function') {
  config = new Function (ctx, `return ${config}`)
}
```

inside postcss-load-config does the job. So when the config is an object e.g loaded from postcssrc.json, package.json, postcss.config.js (module.exports = {...config}). It will be possible to pass context to the config either from the loader ```this.addDependncyTo()``` or use a function within postcss.config.js from the start

**Comment by ai** [comment]
@michael-ciniawsky awesome!

@ertrzyiks you will kill me :D sorry, for real big delay. I will try to release it tomorrow. Feel free to ping me every day since Saturday.

(The only excuse for me, that I did awesome new secret open source :) )

**Comment by ertrzyiks** [comment]
@ai no worries, it's not any urgent to me and secret open source sounds like a perfect excuse!

**Comment by michael-ciniawsky** [comment]
@ai yep no need to rush, we need to stablize config before anyways :). When this PR gets merge interested in the 'offical' loader README boilerplate I made for loaders? [README](https://github.com/michael-ciniawsky/README).

Off topic here but should I open respective issues @postcss-cli @gulp-postcss for PR discussion first? E.g yesterday I updated [gulp-posthtml](https://github.com/posthtml/gulp-posthtml) and woud like to update the tests in the same direction (official gulp guidlines) for gulp-postcss aswell + make a few small refactorings despite adding load-config. Same for CLI.

**Comment by ai** [comment]
Finally merged. Thanks, great work. I just need to update docs, add `result.messages` with `dependencies` and we are ready for release.

**Comment by michael-ciniawsky** [comment]
@ai [docs](https://github.com/michael-ciniawsky/postcss-loader/tree/readme) WIP :)

Sponsor badge directly under the logo or is the sponors section at the bottom ok aswell?

postcss-load-config supports passing a function now, finishing the eval stuff and test it with especially this loader, if it does what it is supposed to be :). @ai can you give me an example/explanation ( little more in-depth :) ) of how the `result.messages.push(dependencies)` will look like?

**Comment by ai** [comment]
@michael-ciniawsky great work! Few moments:

1. Yeap, sponsor badge should be right after main description.
2. You could remove my name from README :).
3. I love your style. But could we keep old, please? =^_^= I love that right now all PostCSS project has same style. Maybe in the future I will be ready for global changes.
4. We also added `plugins` loader query params for webpack 2.0 https://github.com/postcss/postcss-loader/pull/104

I could pull your changes and fix all this list tomorrow (you did a lot with config loader) :).

**Comment by ai** [comment]
Hm, by the way, maybe somebody want to take maintaining for this loader? It is not hard.

But I ma not webpack user at all. You showed great productivity and good understanding of webpack.

I think project will be better if maintainer could at least test it on work project :D.

**Comment by michael-ciniawsky** [comment]
@ai 

1. kk
2. you mean change to @ai, or completely, guessing the latter 😛 
3. yep of course, didn't know there was already consitent styleing accress postcss org, 

> Maybe in the future I will be ready for global changes.

 I can do and would liveto do it, but secondary chore work :)

4. gotcha i will add it.

I can do it, already maintaining html-loader, posthtml-loader. webpack org peasant member :D
Maybe @ertrzyiks wants to give a hand too, we can share the 'burden'

Send invite and grant access here, I can also transfer the load-* stuff then and update it at the weekend, to finally move out of alpha, beta. It makes things a lot easier for me (update it to v1.0.0 asap add jsdoc, keep track of webpack 2 beta etc.)

**Comment by michael-ciniawsky** [comment]
👍 

But transfering the postcss-load-* repos doesn't work, should the stay at my github?

**Comment by ai** [comment]
@michael-ciniawsky yeap, you could keep it :)

**Comment by michael-ciniawsky** [comment]
@ai Can some the content related ajustments stay?

> yeap, you could keep it :)

kk

**Comment by ai** [comment]
@michael-ciniawsky what exactly “content related ajustments”?

I like that you split webpack 1 and webpack 2. But maybe we should promote your config first?

1. Install
2. Usage
  1. PostCSS Config
  2. Webpack 2 config
  3. Webpack 3 config
3. Options
4. Examples

**Comment by michael-ciniawsky** [comment]
@ai  [README](https://github.com/michael-ciniawsky/postcss-loader/tree/readme)

The structure is awesome :+1:
I had the example when load-config is finished + #104

**Comment by ai** [comment]
@michael-ciniawsky I pulled docs changes, thanks.

**Comment by michael-ciniawsky** [comment]
@ai so you want to push them later? I'm asking because otherwise I will do it today, during the day with postcss-load-config release? :) It's neglectible, adding the postcss.config.js example later, just don't  want to mess things up here

**Comment by ai** [comment]
I wrk polish docs today 😊👍

**Comment by michael-ciniawsky** [comment]
👍

**Comment by ai** [comment]
I finished all changes and ready for release. Could somebody test this plugin on real application (just replace version in `package.json` to `postcss/postcss-loader`).

**Comment by richardscarrott** [comment]
@ai I get an webpack validation error when using the following syntax:

```
{
        loader: 'postcss-loader',
        plugins: () => [require('autoprefixer')]
}
```

I think Webpack 2 expects loaders to nest options like this, e.g.

```
{
        loader: 'postcss-loader',
        options: {
            plugins: () => [require('autoprefixer')]
        }
}
```

webpack: 2.1.0-beta.25
postcss-loader:  postcss/postcss-loader (d867d02c6a5af853cde13b3ea09b04c17ae08d41)

**Comment by michael-ciniawsky** [comment]
@richardscarrott yep, i was wondering about that too, your example with 

```js
{
  loader: 'postcss-loader'
  options: {
    plugins: () => [ require('autoprefixer') ]
  }
}
```

is the correct one, but to my knowledge, it's currently not possible to use a function inside `options`. This is a work in progress and currently 'polyfilled' by the **LoaderOptionsPlugin**

```js
const { LoaderOptionsPlugin } = require('webpack')

const config = {
  module: {
   rules: [
   {
     test: /\.css$/,
     use: [ 'style-loader', {loader: 'css-loader', options: { importLoaders: 1 }}, 'postcss-loader'  ]
    }
  },
  plugins : [
    new LoaderOptionsPlugin({
      options: {
        postcss: () => {
          return [ require('autoprefixer') ]
       }
     }
   })
 ]
}

module.exports = config
```

I check the webpack slack quickly, just one moment please :)

**Comment by ai** [comment]
@richardscarrott @michael-ciniawsky fix for webpack 2 config example: e6351b8

Is it correct?

**Comment by richardscarrott** [comment]
@michael-ciniawsky yeah I just noticed it does work with `options.plugins` as a function but as soon as you try to use the ExtractText plugin it fails because that serializes the query / options.

Looks like your LoaderOptionsPlugin workaround is the best approach or I guess create a separate postcss.config.js.

@ai I can't find that commit?

**Comment by michael-ciniawsky** [comment]
@richardscarrott kk, I will check that 
@ai can't find that commit too 😛 , by the way an example with ExtractTextPlugin maybe valuable in general :)

**Comment by ai** [comment]
Oops, it was merge confict :D. Fixed.

**Comment by michael-ciniawsky** [comment]
@ai  basically yes, currently nope 😛 , and combined with `ExtractTextPlugin` for now `LoaderOptionsPlugin` is definitely needed, I add an example for usage with `ExtractTextPlugin`, but let me finished all the postcss-load-config stuff first :)

```js
{
   loader: 'postcss-loader'
   options: { 
     plugins: () => [...plugins] 
   } // Ok, but for plugins only
}
```

```js
{
   loader: 'postcss-loader'
   options: { parser: () => sugarss , plugins: () => [...plugins] } // Error
}
```

**!==**

```js
{
   loader: 'postcss-loader'
   options: () => { parser: sugarss, plugins: [...plugins] } 
   // According to @sokra this should be possible since webpack v2.1.0-beta.25, 
   // when ['loader-utils'].parseQuery(loader.query) is used as it currently is in postcss-loader
   // but API Schema complains that typeof module.rules[1].use[2].options === 'string' || 'options'
   // always fun, fun... , fun...... :D  
}
```

But there seems to be an issue with parseOptions in postcss-loader in general

**webpack.config.js**
**1.**
```js
{
   loader: 'postcss-loader',
   options: {
     plugins: () => [ require('postcss-nested') ]
   }
}
```

 **postcss-loader/index.js**
```js
var params = loaderUtils.parseQuery(loader.query)

console.log(params)
// => {} params.plugins stripped
```

**2.**
```js
{
   loader: 'postcss-loader',
   options: {
     parser: 'sugarss',
     plugins: () => [ require('postcss-nested') ]
   }
}
```

**postcss-loader/index.js**
```js
var params = loaderUtils.parseQuery(loader.query)

console.log(params)
// => { parser: 'sugarss' } params.plugins stripped
```

**3.**
```js
{
   loader: 'postcss-loader',
   options: {
     parser: 'sugarss',
     plugins: [ require('postcss-nested') ]
   }
}
```

**postcss-loader/index.js**
```js
var params = loaderUtils.parseQuery(loader.query)

console.log(params)
// => { parser: 'sugarss' plugins: [ null ] } params.plugins not loaded
```

**Comment by ai** [comment]
@MoOx @michael-ciniawsky so in practice webpack anyway strip functions from options?

So, should we remove `plugins` from query options and force to use `postcss.config.js`?

**Comment by MoOx** [comment]
Webpack 2 remove this restriction. You can pass anything in query now.

**Comment by ai** [comment]
@MoOd it is great. But @michael-ciniawsky shows different 😕. Do we have any mistakes? Maybe we should replace `loaderUtils.parseQuery` to accepts functions?

**Comment by ai** [comment]
I made a test case: https://github.com/ai/webpack2-postcss-test

`plugins` key works perfect from webpack config, but postcss config has same issue. I will open a issue in postcss-load-config.

**Comment by MoOx** [comment]
It's written in https://github.com/webpack/webpack/releases/tag/v2.1.0-beta.23
If this does not work, you are doing something wrong.

**Pull request merged**